store/tikv/gc_worker.go

```diff
@@ -36,17 +36,48 @@ import (
 
 // GCWorker periodically triggers GC process on tikv server.
 type GCWorker struct {
-	uuid        string
-	desc        string
-	store       *tikvStore
-	gcIsRunning int32
+	uuid  string
+	desc  string
+	store *tikvStore
+
+	/*
+	 * gc worker uses "session"
+	 * safepoint checker uses "spSession"
+	 * so there is no race
+	 */
+	session   tidb.Session
+	spSession tidb.Session
+
+	gcIsRunning bool
 	lastFinish  time.Time
 	cancel      goctx.CancelFunc
 	done        chan error
 }
 
+// StartSafePointChecker triggers SafePoint Checker on each tidb
+func (w *GCWorker) StartSafePointChecker() {
+	for {
+		select {
+		case <-w.store.spMsg:
+			return
+		default:
+			spCachedTime := time.Now()
+			cachedSafePoint, cacheErr := w.loadSafePoint(gcSavedSafePoint)
+			if cacheErr == nil {
+				w.store.spMutex.Lock()
+				w.store.safePoint = cachedSafePoint
+				w.store.spTime = spCachedTime
+				w.store.spMutex.Unlock()
+			} else {
+				log.Error("[safepoint] read error:", cacheErr)
+			}
+			time.Sleep(5 * time.Second) // this is configurable
```

Comment:
You should write code like this:
```
ticker := time.NewTicker()
select {
    case <-w.store.spMsg:
    case <-ticker.C:
        ...
}
```
Because when you `Sleep()`, w.store.spMsg signal will be miss.